### PR TITLE
[SEDONA-35] address user-data mutability issue with `Adapter.toDF()`

### PR DESCRIPTION
**Pull request by yitao-li**
Signed-off-by: Yitao Li <yitao@rstudio.com>

## Is this PR related to a proposed Issue?

https://issues.apache.org/jira/browse/SEDONA-35

I think we should create an issue for it.

The problem is currently if `Adapter.toDf()` is called with a `SpatialRDD` carrying non-spatial attributes (a.k.a "user data") in addition to spatial attributes, then the Spark dataframe returned from `toDf()` can only be evaluated successfully once. If it was  evaluated more than once, then a null pointer exception occurrs. If a Spark dataframe is not cached, then the underlying RDD could be evaluated more than once in many scenarios, so, the NPE problem could break quite a few use cases.

## What changes were proposed in this PR?

Essentially, the changes in this PR ensures `toDf()` respects the immutability of `RDD`s in Spark. The abovementioned NPE issue happens because the `Geometry` objects stored in the input `RDD` have their `userData` set to `null` within a `RDD.map` transformation, and this causes unexpected `null` values (at least when Spark is running in "local" mode) when this transformation is evaluated the second time.

So, the short-term fix (as shown in this PR) is to operate on a copy of the `Geometry` object, rather than modifying the `Geometry` object that is already stored in the `RDD`.

Long-term fix: I think it would be a better practice to work with `RDD[ImmutableGeometry]`s where `ImmutableGeometry` is a sub-class of `Geometry` that does not permit modifications of its existing data fields, but permits copying its underling data to a mutable `Geometry` object. I'm unsure whether there may be noticeable perf impact with this type of change though and I hope to get a second opinion on this. But this feels like, in principle, the correct thing to do, because `RDD`s are by design supposed to be immutable in Spark, and correctness issue arises if the immutability contract is not followed.

## How was this patch tested?

Unit test (in progress)

## Did this PR include necessary documentation updates?

Not applicable -- there is no change to documentation

### CODE REVIEW

**Comment by jiayuasu** [comment]
@yitao-li This looks good to me. I actually ran into this issue before but wasn't able to identify the issue. Thanks for pointing this out. If you can add some test cases, I will accept this PR immediately.

ImmutableGeometry may not be a good choice because many geoms in JTS directly extend Geometry. If we add a ImmutableGeometry, we have to develop corresponding wrappers for all different geom types, which is not practical I believe

File: sql/src/main/scala/org/apache/sedona/sql/utils/Adapter.scala [hunk 1]
```diff
@@ -166,8 +169,9 @@ object Adapter {
     if (fieldNames != null && fieldNames.nonEmpty) {
       val userData = "" + geom.getUserData.asInstanceOf[String]
       val fields = userData.split("\t")
-//      geom.setUserData(null) // Set to null will lead to the null pointer exception of the previous line. Not sure why.
-      (Seq(geom), fields)
```
Comment (yitao-li) [inline sql/src/main/scala/org/apache/sedona/sql/utils/Adapter.scala:170]: I guess while we are at it, we should also address this issue here as well (even though it's not as bad as a null pointer exception) -- also, see test case with pairrdd that should cover this code path

File: sql/src/test/scala/org/apache/sedona/sql/adapterTestScala.scala [hunk 2]
```diff
@@ -198,5 +199,51 @@ class adapterTestScala extends TestBaseScala with Matchers with GivenWhenThen{
       val spatialDf = Adapter.toDf(spatialRDD, sparkSession)
       assert(spatialDf.schema.fields(1).name == "LST")
     }
+
+    it("can convert spatial RDD with user data to a valid Dataframe") {
```
Comment (yitao-li) [inline sql/src/test/scala/org/apache/sedona/sql/adapterTestScala.scala:203]: I think this should cover the case of spatial RDD -> dataframe

File: sql/src/test/scala/org/apache/sedona/sql/adapterTestScala.scala [hunk 3]
```diff
@@ -198,5 +199,51 @@ class adapterTestScala extends TestBaseScala with Matchers with GivenWhenThen{
...
+        assert(rows.length == 1)
+        val geom = rows(0).get(0).asInstanceOf[Point]
+        assert(geom.getX == 40.7128)
+        assert(geom.getY == -74.006)
+        assert(rows(0).get(1).asInstanceOf[String] == "attr1")
+        assert(rows(0).get(2).asInstanceOf[String] == "attr2")
+      }
+    }
+
+    it("can convert spatial pair RDD with user data to a valid Dataframe") {
```
Comment (yitao-li) [inline sql/src/test/scala/org/apache/sedona/sql/adapterTestScala.scala:220]: I think this should cover the case of pair RDD -> dataframe

File: sql/src/main/scala/org/apache/sedona/sql/utils/Adapter.scala [hunk 4]
```diff
@@ -166,8 +169,9 @@ object Adapter {
...
+      val geomWithoutUserData = geom.copy
+      geomWithoutUserData.setUserData(null)
+      (Seq(geomWithoutUserData), fields)
```
Comment (yitao-li) [inline sql/src/main/scala/org/apache/sedona/sql/utils/Adapter.scala:174]: I guess while we are at it, we should also address this issue here as well (even though it's not as bad as a null pointer exception) -- also, see test case with pairrdd that should cover this code path

**Comment by yitao-li** [comment]
@jiayuasu Thanks for the feedback :slightly_smiling_face:

I had a chat with one of my colleagues about this as well, and we were both thinking if Java or Scala had the equivalent of the `const` qualifier in C++, then there would have been a simpler and more concise way to specify immutability requirements. But because such language construct does not exist in Java/Scala yet, we'll have to go with the current solution.

Also, I fixed a similar (but different) issue with PairRDD and created some test cases covering both code paths.

**Review by jiayuasu — APPROVED** [review]

**Comment by jiayuasu** [comment]
@yitao-li Thanks for this PR. I will merge it and enter the release process of Sedona 1.0.1 which is mainly for bug fix purpose
Then I will get back to your R Interface PR.

**Pull request merged**